File: java/src/Stack.java
import java.util.Iterator;
import java.util.NoSuchElementException;

public class Stack<T> implements Iterable<T> {

    private LinkedList<T> head;

    public void push(T x) {
        head = new LinkedList<>(x, null, head);
    }

    public T pop() {
        if (head == null)
            throw new NoSuchElementException();

        var value = head.value;
        head = head.next;

        return value;
    }

    public Iterator<T> iterator() {
        return new LinkedListIterator<>(head);
    }
}

File: c/stack.c
#include "linked_list.c"
#include <stdlib.h>

typedef struct {
  linked_list *head;
} stack;

void push(stack *s, int x) {
  linked_list *new_head = malloc(sizeof(linked_list));

  new_head->next = s->head;
  new_head->value = x;

  s->head = new_head;
}

int pop(stack *s) {
  int value = s->head->value;

  linked_list *new_head = s->head->next;

  free(s->head);

  s->head = new_head;

  return value;
}
